AP_Scheduler: corrected tick counter overflow handling, fixes #17642

ntamas opened the pull request:
This PR fixes the issue outlined with the task scheduler in #17642.

Right now the scheduler will prematurely schedule all tasks at the same time every 65536 ticks (~163 seconds on a 400 Hz loop). The PR fixes the problem by reverting the type of the tick count difference variables to `uint16_t` from `uint32_t`.

A more detailed description of the problem (along with symptoms) is in #17642.

peterbarker commented on libraries/AP_Scheduler/AP_Scheduler.cpp:
```diff
@@ -160,9 +160,9 @@ void AP_Scheduler::run(uint32_t time_available)
         const AP_Scheduler::Task& task = (i < _num_unshared_tasks) ? _tasks[i] : _common_tasks[i - _num_unshared_tasks];
 
-        uint32_t dt = _tick_counter - _last_run[i];
+        uint16_t dt = _tick_counter - _last_run[i];
```
```suggestion
        const uint16_t dt = _tick_counter - _last_run[i];
```

peterbarker commented on libraries/AP_Scheduler/AP_Scheduler.cpp:
```diff
@@ -160,9 +160,9 @@ void AP_Scheduler::run(uint32_t time_available)
+        uint16_t dt = _tick_counter - _last_run[i];
         // we allow 0 to mean loop rate
-        uint32_t interval_ticks = (is_zero(task.rate_hz) ? 1 : _loop_rate_hz / task.rate_hz);
+        uint16_t interval_ticks = (is_zero(task.rate_hz) ? 1 : _loop_rate_hz / task.rate_hz);
```
Pretty sure this patch isn't required.  Someone running their loops *much* faster than the slowest task might end up getting problems in here.  e.g. `_loop_rate_hz` of 1000 and a `task.rate_hz` of 1/60 gives an interval_ticks of 60,000 - perilously close the uint16-max.  I think we can afford to burn an extra 2 bytes on the stack here :-)

ntamas replied:
I wasn't sure about this change either; I changed the `uint32_t` to `uint16_t` for sake of consistency with `dt`.

One thing is sure: with the current scheduler architecture, we cannot schedule tasks with an interval larger than 65535 ticks anyway because the tick counter is 16-bit. Therefore,

* If we keep `interval_ticks` as `uint32_t`, it will always be larger than the maximum possible value of `dt` for such tasks if `dt` is 16-bit, so tasks would never get scheduled.
* If we convert `interval_ticks` to `uint16_t`, it could potentially overflow for such rarely-scheduled tasks, and it could potentially overflow in a way that the task gets scheduled much more frequently than expected.

So, it seems like `uint32_t` is safer as we don't cause extra CPU load even if the user misconfigures the loop.

tridge commented:
@ntamas thanks!

peterbarker reviewed (dismissed):
Sorry, we need to squash all those commits together.  Just fixup the second two into the first.

Tried to it for you, but permission was denied on your branch.

ntamas commented:
Not sure why it was denied on my branch or where I could enable it, but I've squashed the commits and rebased it on top of master so I hope it's good to go now.

peterbarker commented:
Merged, thanks!

In terms of the fail-to-push thing, that was perhaps (a) me screwing up or (b) you unchecked the box allowing pushes from maintainers to your branch.  I think it can also happen with private repos or something in github.

In any case, thanks for this great fix!

ntamas commented:
Figured it out: the checkbox that allows pushes for maintainers is apparently not visible if the branch being pulled from belongs to an organization account. The checkbox is there if I make a PR from my own account but not from an organization account.

Anyhow, thanks for merging this!
